File: Codes/Hall1997.py
from numpy import exp, sqrt, pi
import numpy as np
from scipy.special import logsumexp


def normpdf(x):
    return 1 / sqrt(2 * pi) * exp(-(x**2) / 2)


def density_upper(t, mu, a, b, x0, trunc_num=100):
    """
    First passage time density on the upper boundary
    """
    tau = mu + b
    a1 = a - x0
    factor = (
        t ** (-1.5)
        * exp(a1 * tau - 0.5 * tau**2 * t - b / (2 * a) * a1**2)
        / sqrt(2 * pi)
    )
    result = 0
    for j in range(trunc_num):
        rj = (2 * j + 1) * a - (-1) ** j * x0
        term = (-1) ** j * rj * exp(0.5 * (b / a - 1 / t) * rj**2)
        if np.max(np.abs(term)) < 1e-20:
            break
        result += term
    return result * factor


def density_lower(t, mu, a, b, x0, trunc_num=100):
    """
    First passage time density on the lower boundary
    """
    tau = -mu + b
    a1 = a + x0
    factor = (
        t ** (-1.5)
        * exp(a1 * tau - 0.5 * tau**2 * t - b / (2 * a) * a1**2)
        / sqrt(2 * pi)
    )
    result = 0
    for j in range(trunc_num):
        rj = (2 * j + 1) * a + (-1) ** j * x0
        term = (-1) ** j * rj * exp(0.5 * (b / a - 1 / t) * rj**2)
        if np.max(np.abs(term)) < 1e-20:
            break
        result += term
    return result * factor


def density_vertical(x, mu, a, b, x0, T, trunc_num=100, if_logsumexp=True):
    """
    doesn't work, haven't figured out why
    """
    x = x - x0
    factor = exp(mu * x - 0.5 * mu**2 * T) / sqrt(T)
    result = normpdf(x / sqrt(T))
    for j in range(1, trunc_num):
        t1 = 4 * b * j * (2 * a * j + x0) - (x - 4 * a * j) ** 2 / (2 * T)
        t2 = 4 * b * j * (2 * a * j - x0) - (x + 4 * a * j) ** 2 / (2 * T)
        t3 = 2 * b * (2 * j - 1) * (2 * a * j - a + x0) - (x + (4 * j - 2) * a + 2 * x0) ** 2 / (2 * T)
        t4 = 2 * b * (2 * j - 1) * (2 * a * j - a - x0) - (x - (4 * j - 2) * a + 2 * x0) ** 2 / (2 * T)
        if if_logsumexp:
            logterm, sign = logsumexp([t1, t2, t3, t4], b=[1, 1, -1, -1], return_sign=True)
            if logterm < -20:
                break
            result += sign * exp(logterm) / sqrt(2 * pi)
        else:
            term = exp(t1) + exp(t2) - exp(t3) - exp(t4)
            if np.max(np.abs(term)) < 1e-50:
                break
            result += term / sqrt(2 * pi)
    return result * factor

File: hall1997_cpp/test/hall1997.cpp
#include <iostream>
#include <cmath>
#include <vector>
#include <chrono>

enum sign
{
    upper = 1,
    lower = -1
};

double fpt_density(double t, double mu, double a, double b, double x0, sign bdy, int trunc_num = 100)
{
    double tau = bdy * mu + b;
    double a1 = a - bdy * x0;
    double factor = (pow(t, -1.5) * exp(a1 * tau - 0.5 * pow(tau, 2) * t - b / (2 * a) * pow(a1, 2)) / sqrt(2 * M_PI));
    double result = 0;
    for (int j = 0; j < trunc_num; j++)
    {
        double rj = (2 * j + 1) * a - bdy * pow(-1, j) * x0;
        double term = pow(-1, j) * rj * exp(0.5 * (b / a - 1 / t) * pow(rj, 2));
        if (fabs(term) < 1e-20)
        {
            break;
        }
        result += term;
    }
    return result * factor;
}

double trans_density(double x, double mu, double a, double b, double x0, double T, int trunc_num = 100)
{
    x = x - x0;
    double factor = exp(mu * x - 0.5 * mu * mu * T) / sqrt(T);
    double result = exp(-0.5 * x * x / T) / sqrt(2.0 * M_PI);
    for (int j = 1; j < trunc_num; ++j)
    {
        double t1 = 4.0 * b * j * (2.0 * a * j + x0) - (x - 4.0 * a * j) * (x - 4.0 * a * j) / (2.0 * T);
        double t2 = 4.0 * b * j * (2.0 * a * j - x0) - (x + 4.0 * a * j) * (x + 4.0 * a * j) / (2.0 * T);
        double t3 = 2.0 * b * (2.0 * j - 1) * (2.0 * a * j - a + x0) - (x + (4.0 * j - 2) * a + 2.0 * x0) * (x + (4.0 * j - 2) * a + 2.0 * x0) / (2.0 * T);
        double t4 = 2.0 * b * (2.0 * j - 1) * (2.0 * a * j - a - x0) - (x - (4.0 * j - 2) * a + 2.0 * x0) * (x - (4.0 * j - 2) * a + 2.0 * x0) / (2.0 * T);
        double term = exp(t1) + exp(t2) - exp(t3) - exp(t4);
        if (fabs(term) < 1e-20)
        {
            break;
        }
        result += term / sqrt(2.0 * M_PI);
    }
    return result * factor;
}